x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PushDownJoinPastProject.java

```diff
@@ -0,0 +1,119 @@
+            // child - so pulling out a Project from the left child would require us to also update the StubRelation (and the Aggregate
+            // on top of it)
+            return join;
+        }
+
+        if (join.left() instanceof Project project && join.config().type() == JoinTypes.LEFT) {
+            AttributeMap.Builder<Expression> aliasBuilder = AttributeMap.builder();
+            project.forEachExpression(Alias.class, a -> aliasBuilder.put(a.toAttribute(), a.child()));
+            var aliasesFromProject = aliasBuilder.build();
+
+            // Propagate any renames into the Join, as we will remove the upstream Project.
+            // E.g. `RENAME field AS key | LOOKUP JOIN idx ON key` -> `LOOKUP JOIN idx ON field | ...`
+            Join updatedJoin = PushDownUtils.resolveRenamesFromMap(join, aliasesFromProject);
+
+            // Construct the expressions for the new downstream Project using the Join's output.
+            // We need to carry over RENAMEs/aliases from the original upstream Project.
+            List<Attribute> originalOutput = join.output();
+            List<NamedExpression> newProjections = new ArrayList<>(originalOutput.size());
+            for (Attribute attr : originalOutput) {
+                Attribute resolved = (Attribute) aliasesFromProject.resolve(attr, attr);
+                if (attr.semanticEquals(resolved)) {
+                    newProjections.add(attr);
+                } else {
+                    Alias renamed = new Alias(attr.source(), attr.name(), resolved, attr.id(), attr.synthetic());
+                    newProjections.add(renamed);
+                }
+            }
+
+            // This doesn't deal with name conflicts yet. Any name shadowed by a lookup field from the `LOOKUP JOIN` could still have been
+            // used in the original Project; any such conflict needs to be resolved by copying the attribute under a temporary name via an
+            // Eval - and using the attribute from said Eval in the new downstream Project.
+            Set<String> lookupFieldNames = new HashSet<>(Expressions.names(join.rightOutputFields()));
+            List<NamedExpression> finalProjections = new ArrayList<>(newProjections.size());
+            AttributeMap.Builder<Alias> aliasesForReplacedAttributesBuilder = AttributeMap.builder();
+            AttributeSet leftOutput = project.child().outputSet();
+
+            for (NamedExpression proj : newProjections) {
+                // TODO: add assert to Project that ensures Alias to attr or pure attr.
+                Attribute coreAttr = (Attribute) (proj instanceof Alias as ? as.child() : proj);
+                // Only fields from the left need to be protected from conflicts - because fields from the right shadow them.
+                if (leftOutput.contains(coreAttr) == false || lookupFieldNames.contains(coreAttr.name()) == false) {
```

Comment:
Could the branching be rewritten to better show the naming "conflict"? I.e. 
```suggestion
                if (leftOutput.contains(coreAttr) && lookupFieldNames.contains(coreAttr.name())) {
```
and have under this condition the conflict resolution?